File: typescript/src/pages/duckduckgo.page.ts
import { BasePage } from './base.page';

/**
 * DuckDuckGo search page. Used instead of Google: Google serves reCAPTCHA
 * bot-challenges to datacenter IPs (see the Java twin's SearchTest notes).
 */
export class DuckDuckGoPage extends BasePage {
  // DDG renders its search field as a TEXTAREA (role=combobox), not an <input>.
  private readonly searchBox = this.page
    .locator("#searchbox_homepage textarea, textarea[role='combobox'], input[name='q']")
    .first();

  async open(url: string): Promise<void> {
    await this.page.goto(url);
    await this.searchBox.waitFor();
  }

  async isLoaded(): Promise<boolean> {
    return this.searchBox.isVisible();
  }

  async searchFor(term: string): Promise<void> {
    await this.searchBox.fill(term);
    await this.searchBox.press('Enter');
  }

  /** Count rendered result titles (waits for the first so the count is stable). */
  async resultCount(): Promise<number> {
    await this.resultTitles().first().waitFor();
    return this.resultTitles().count();
  }

  /**
   * Open the first result whose title mentions the given text (e.g. "IMDb").
   * DDG wraps result links in redirect URLs, so match on title text, not href.
   */
  async openResultMentioning(text: string): Promise<void> {
    const result = this.resultTitles().filter({ hasText: text }).first();
    await result.scrollIntoViewIfNeeded();
    await result.click();
  }

  private resultTitles() {
    return this.page.locator("a[data-testid='result-title-a'], a.result__a");
  }
}

File: java/src/main/java/io/github/mb1lal/playwright/ui/pages/DuckDuckGoPage.java
package io.github.mb1lal.playwright.ui.pages;

import io.github.mb1lal.playwright.base.TestContext;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import lombok.extern.slf4j.Slf4j;

/**
 * DuckDuckGo search page.
 *
 * <p>Used instead of Google for the search demos: Google serves reCAPTCHA
 * bot-challenges to datacenter IPs, while DuckDuckGo is automation-friendly
 * and has no blocking consent dialog.
 */
@Slf4j
public class DuckDuckGoPage extends BasePage {

    private final Locator searchBox;

    public DuckDuckGoPage(Page page, TestContext testContext) {
        super(page, testContext);
        // NOTE: DDG renders its search field as a TEXTAREA (role=combobox),
        // not an <input>. Union selector covers classic and current markup.
        this.searchBox = page.locator(
                "#searchbox_homepage textarea, textarea[role='combobox'], input[name='q']").first();
    }

    public void open(String url) {
        log.info("Opening DuckDuckGo at {}", url);
        page.navigate(url);
        page.waitForLoadState();
        searchBox.waitFor();
    }

    public boolean isLoaded() {
        return isVisible(searchBox, "DuckDuckGo search box");
    }

    /** Search for a term by filling the box and pressing Enter. */
    public void searchFor(String term) {
        log.info("Searching DuckDuckGo for '{}'", term);
        fill(searchBox, term, "DuckDuckGo search box");
        searchBox.press("Enter");
        page.waitForLoadState();
    }

    /**
     * Number of rendered result titles. Waits for the first result so the
     * count is never taken on a half-loaded page.
     */
    public int resultCount() {
        resultTitles().first().waitFor();
        int count = resultTitles().count();
        log.debug("DuckDuckGo result count: {}", count);
        return count;
    }

    /**
     * Open the first result whose title mentions the given text (e.g. "IMDb").
     * DuckDuckGo wraps result links in redirect URLs, so results are matched
     * by title text rather than href.
     */
    public void openResultMentioning(String text) {
        log.info("Opening search result mentioning '{}'", text);
        Locator result = resultTitles()
                .filter(new Locator.FilterOptions().setHasText(text)).first();
        scrollIntoView(result, "Search result mentioning '" + text + "'");
        click(result, "Search result mentioning '" + text + "'");
        page.waitForLoadState();
    }

    private Locator resultTitles() {
        return page.locator("a[data-testid='result-title-a'], a.result__a");
    }
}
